File: src/main/java/com/mallang/auth/infrastructure/oauth/github/GithubAuthCodeRequestUrlProvider.java
package com.mallang.auth.infrastructure.oauth.github;

import com.mallang.auth.domain.OauthId.OauthServerType;
import com.mallang.auth.domain.oauth.AuthCodeRequestUrlProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.web.util.UriComponentsBuilder;

@RequiredArgsConstructor
@Component
public class GithubAuthCodeRequestUrlProvider implements AuthCodeRequestUrlProvider {

    private final GithubOauthConfig githubOauthConfig;

    @Override
    public OauthServerType supportServer() {
        return OauthServerType.GITHUB;
    }

    @Override
    public String provide() {
        return UriComponentsBuilder
                .fromUriString("https://github.com/login/oauth/authorize")
                .queryParam("response_type", "code")
                .queryParam("client_id", githubOauthConfig.clientId())
                .queryParam("redirect_uri", githubOauthConfig.redirectUri())
                .queryParam("scope", String.join(",", githubOauthConfig.scope()))
                .toUriString();
    }
}

File: src/main/kotlin/com/mallang/auth/infrastructure/oauth/github/GithubAuthCodeRequestUrlProvider.kt
package com.mallang.auth.infrastructure.oauth.github

import com.mallang.auth.domain.OauthServerType
import com.mallang.auth.domain.OauthServerType.GITHUB
import com.mallang.auth.domain.oauth.AuthCodeRequestUrlProvider
import org.springframework.stereotype.Component
import org.springframework.web.util.UriComponentsBuilder

@Component
class GithubAuthCodeRequestUrlProvider(
        private val githubOauthConfig: GithubOauthConfig
) : AuthCodeRequestUrlProvider {

    override fun supportServer(): OauthServerType = GITHUB

    override fun provide(): String =
            UriComponentsBuilder
                    .fromUriString("https://github.com/login/oauth/authorize")
                    .queryParam("response_type", "code")
                    .queryParam("client_id", githubOauthConfig.clientId)
                    .queryParam("redirect_uri", githubOauthConfig.redirectUri)
                    .queryParam("scope", githubOauthConfig.scope.joinToString(","))
                    .toUriString()
}
